### PR TITLE
Fix #80, support an include mechanism for PDI config

### LINKED ISSUE #80: Support an include mechanism for PDI config
In GitLab by @jbigot on Mar 23, 2018, 12:00

### PR DESCRIPTION
**Pull request by JAuriac**
Support an include mechanism for PDI config, to include other YAML files into root YAML document.
Depends on Paraconf handling of relative path to point to included YAML files.

# List of things to check before making a PR

Before merging your code, please check the following:

* [x] you have added a line describing your changes to the Changelog;
* [x] you have added unit tests for any new or improved feature;
* [x] in case you updated dependencies, you have checked pdi/docs/CheckList.md;
* [x] in case of a change in pdi.h, this same change must be reflected in no-pdi/include/pdi.h;
* you have checked your code format:
  - [ ] you have checked that you respect all conventions specified in CONTRIBUTING.md;
  - [x] you have checked that the indentation and formatting conforms to the `.clang-format`;
  - [ ] you have documented with doxygen any new or changed function / class;
* you have correctly updated the copyright headers:
  - [x] your institution is in the copyright header of every file you (substantially) modified;
  - [x] you have checked that the end-year of the copyright there is the current one;
* you have updated the AUTHORS file:
  - [x] you have added yourself to the AUTHORS file;
  - [x] if this is a new contribution, you have added it to the AUTHORS file;
* you have added everything to the user documentation:
  - [ ] any new CMake configuration option;
  - [x] any change in the yaml config;
  - [ ] any change to the public or plugin API;
  - [ ] any other new or changed user-facing feature;
  - [ ] any change to the dependencies;
* you have correctly linked your MR to one or more issues:
  - [x] your MR solves an identified issue;
  - [x] your commit contain the `Fix #issue` keyword to autoclose the issue when merged.

### CODE REVIEW

File: pdi/src/global_context.cxx [hunk 1]
```diff
@@ -102,37 +102,55 @@ void Global_context::finalize()
...
+	if (!PC_status(pdi)) {
+		root = pdi;
+	}
+
+	PC_tree_t includes = PC_get(root, ".include");
+	if (!PC_status(includes)) {
+		PDI::each(includes, [&](PC_tree_t yaml_subfile){ load_pdi_config(ctx, PC_parse_path((PDI::to_string(yaml_subfile)).c_str())); });
+	} else {
+		ctx.logger().trace("Includes are not defined in specification tree");
+	}
```
Comment (Yushan-Wang) [inline pdi/src/global_context.cxx]: no need for `else`

File: tests/test_07.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+pdi:
+  include:
+    - /<full>/<path>/<to>/test_07_data.yml
```
Comment (Yushan-Wang) [inline tests/test_07.yml]: How does CI fill out the full path?

File: tests/test_07_data.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+pdi:
+  logging: trace
```
Comment (Yushan-Wang) [inline tests/test_07_data.yml]: Is the `logging` parsed here or in the other yaml?

File: tests/test_07.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  logging: trace
+  metadata:
```
Comment (Yushan-Wang) [inline tests/test_07.yml]: Could you add a `data` here as well, to test the case where we have duplicates?

File: tests/test_07_data.yml (hunk 3 above)
Comment (JAuriac) [reply tests/test_07_data.yml]: The logging is parsed like other variables, so the latest yaml read overwrites, so the logging used is the one specified in test_07.yaml (we read the included yaml before the root yaml)

File: tests/test_07.yml (hunk 4 above)
Comment (JAuriac) [reply tests/test_07.yml]: Done
This data is parsed like other variables, so the latest yaml read overwrites, so if a data is set to an int in the root yaml, it will take precedence (we read the included yaml before the root yaml)

File: tests/test_07.yml (hunk 2 above)
Comment (JAuriac) [reply tests/test_07.yml]: I can set this path to the one used by the github runner, but the local ctest will fail

File: tests/test_07.yml (hunk 4 above)
Comment (Yushan-Wang) [reply tests/test_07.yml]: It should throw an error: can not redefine the same variable

File: pdi/src/global_context.cxx [hunk 5]
```diff
@@ -102,37 +102,53 @@ void Global_context::finalize()
 	s_context.reset();
 }
 
+void load_pdi_config(Global_context& ctx, PC_tree_t conf)
```
Comment (Yushan-Wang) [inline pdi/src/global_context.cxx]: In this function, I don't see any parsing beyond `.types`, `.data`, and `metadata`. So the `logging` defined in the subyaml will not be parsed.

Comment (JAuriac) [reply pdi/src/global_context.cxx]: Indeed, we are keeping the previous signature of Global_context:
```
Global_context::Global_context(PC_tree_t conf)
	: m_logger{"PDI", PC_get(conf, ".logging")}
	, m_plugins{*this, conf}
	, m_callbacks{*this}
```
So the `.logging` is only read in the root yaml.

File: tests/test_07_data.yml (hunk 3 above)
Comment (JAuriac) [reply tests/test_07_data.yml]: We are keeping the previous signature of Global_context:
```
Global_context::Global_context(PC_tree_t conf)
	: m_logger{"PDI", PC_get(conf, ".logging")}
	, m_plugins{*this, conf}
	, m_callbacks{*this}
```
So the `.logging` is only read in the root yaml, contrarily to the other members of the configuration.

File: tests/test_07_data.yml [hunk 6]
```diff
@@ -0,0 +1,21 @@
+pdi:
+  logging: debug
```
Comment (Yushan-Wang) [inline tests/test_07_data.yml]: remove `logging`

File: pdi/src/global_context.cxx [hunk 7]
```diff
@@ -102,37 +102,53 @@ void Global_context::finalize()
...
+{
+	PC_tree_t root = conf;
+	// if we have a node ".pdi", we go down inside it, otherwise we stay at this level
+	// this is used for both for the root yaml and for included yaml
+	PC_tree_t pdi = PC_get(conf, ".pdi");
+	if (!PC_status(pdi)) {
+		root = pdi;
+	}
```
Comment (jbigot) [inline pdi/src/global_context.cxx]: please don't do that ! It's up to the user to do that if they want.

Maybe the syntax could support 2 variants:
either:
```yaml
include:
- /path/to/my/file.yaml
```
or
```yaml
include:
- file: /path/to/my/file.yaml
  node: .pdi[5]
```
where `node` is any valid ypath (paraconf) expression

**Comment by JAuriac** [comment]
As we currently need the full path to read a yaml subfile, the test_07 works on the github runner for linux (through a workaround overwriting a placeholder field in the root yaml file) but not for macos, and not for local

File: tests/test_07.yml (hunk 4 above)
Comment (JAuriac) [reply tests/test_07.yml]: Done, modified test_07_C, and added both check_duplicate and load_pdi_config_duplicate_data_via_inclusion

Comment (Yushan-Wang) [reply tests/test_07.yml]: Could you make another test where you define the same variable in different yaml files? PDI is expected to throw the error.

**Comment by JAuriac** [comment]
The integrations tests "test_duplicate_yaml_include" and "test_circular_yaml_include" are disabled by default, as they expect failure. We use instead the unitary tests "check_duplicate" and "load_pdi_config_duplicate_data_via_inclusion", using System_error.
This last commit only test using absolute paths for the CI, with Paraconf 1.0.3 (also tested locally for both absolute and relative paths, with Paraconf 1.1.1).

File: spack.yaml [hunk 8]
```diff
@@ -36,6 +36,7 @@ spack:
   - 'mpi'
   - 'netcdf-c@4.8.1:4'
   - 'paraconf@1.0.0:1'
+  # - 'paraconf@1.1.0:1'
```
Comment (Yushan-Wang) [inline spack.yaml]:
```suggestion
```

File: CMakeLists.txt [hunk 9]
```diff
@@ -251,13 +251,15 @@ endif()
 # if ( PDI or ... )
 sbuild_add_dependency(paraconf "${USE_DEFAULT}"
 	EMBEDDED_PATH "vendor/paraconf-1.0.3"
+	# EMBEDDED_PATH "vendor/paraconf-1.1.1"
```
Comment (Yushan-Wang) [inline CMakeLists.txt]:
```suggestion
```

File: CMakeLists.txt [hunk 10]
```diff
@@ -251,13 +251,15 @@ endif()
...
 	COMPONENTS ${PARACONF_COMPONENTS}
 	CMAKE_CACHE_ARGS
 		"-DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=TRUE"
 		"-DUSE_yaml:STRING=SYSTEM"
 		"-DBUILD_TESTING:BOOL=OFF"
 	DEPENDS "yaml"
 	VERSION 1.0
+	# VERSION 1.1
```
Comment (Yushan-Wang) [inline CMakeLists.txt]:
```suggestion
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -28,6 +28,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ### Added
 * Added tests for using an installed PDI
 
+### Changed
+* Update required version of paraconf to 1.1.0
+
```
Comment (Yushan-Wang) [inline CHANGELOG.md]:
```suggestion
```

File: pdi/src/global_context.cxx [hunk 12]
```diff
@@ -57,24 +64,38 @@ using std::unordered_map;
...
+#ifdef PARACONF_VERSION_MAJOR
+#if (PARACONF_VERSION_MAJOR > 1) || (PARACONF_VERSION_MAJOR == 1 && PARACONF_VERSION_MINOR >= 1)
+#define PDI_HAS_PC_PATH 1
+#else
+#define PDI_HAS_PC_PATH 0
+#endif
+#else
+#define PDI_HAS_PC_PATH 0
+#endif
+
```
Comment (Yushan-Wang) [inline pdi/src/global_context.cxx]: is this really necessary?

File: pdi/src/global_context.cxx [hunk 13]
```diff
@@ -57,24 +64,38 @@ using std::unordered_map;
...
 namespace PDI {
 
 namespace {
 
-void load_data(Context& ctx, PC_tree_t node, bool is_metadata)
+void load_data(Global_context& ctx, PC_tree_t node, bool is_metadata)
```
Comment (Yushan-Wang) [inline pdi/src/global_context.cxx:80]: why changing to `global_context`?

File: pdi/src/global_context.cxx [hunk 14]
```diff
@@ -109,31 +130,21 @@ Global_context::Global_context(PC_tree_t conf)
...
 	// load basic datatypes
 	Datatype_template::load_basic_datatypes(*this);
-	// load user datatypes
-	Datatype_template::load_user_datatypes(*this, PC_get(conf, ".types"));
+
+	// Pass 1: collect all plugin declarations from the full include tree,
+	//         then load them - so custom types (like MPI_Comm) are available.
+	{
+		std::unordered_set<std::string> visited;
+		collect_plugins_impl(conf, visited);
```
Comment (Yushan-Wang) [inline pdi/src/global_context.cxx]: why this is required?

File: pdi/src/global_context.cxx (hunk 12 above)
Comment (JAuriac) [reply pdi/src/global_context.cxx]: Yes, to use absolute paths with Paraconf < 1.1, and both absolute and relative paths with Paraconf 1.1+

File: pdi/src/global_context.cxx (hunk 13 above)
Comment (JAuriac) [reply pdi/src/global_context.cxx:80]: Changed back

File: pdi/src/global_context.cxx (hunk 14 above)
Comment (JAuriac) [reply pdi/src/global_context.cxx]: This was used to handle the dual pass. We are now back to a single pass on the PC tree, with two sub-pass to read plugins, then types/metadata/data.

File: pdi/src/global_context.cxx (hunk 12 above)
Comment (jbigot) [reply pdi/src/global_context.cxx]: I'd recommend [this pattern](https://github.com/pdidev/pdi/blob/improved_errmsg/pdi/include/pdi/error.h#L151):

```suggestion
#if defined(PARACONF_VERSION) && (PARACONF_UNTYPED_VERSION >= PARACONF_UNTYPED_COMPUTE_VERSION(1, 1, 0))
#define PDI_HAS_PC_PATH 1
#else
#define PDI_HAS_PC_PATH 0
#endif

```

of you could directly do the test where needed

File: pdi/src/global_context.cxx
Comment (jbigot) [inline pdi/src/global_context.cxx:267]: this will load the plugins at the construction of the object and should be reworked

File: pdi/include/pdi/context.h [hunk 15]
```diff
@@ -135,6 +135,9 @@ class PDI_EXPORT Context
 
 	/// Finalizes PDI and exits application
 	virtual void finalize_and_exit() = 0;
+
+	// Check duplicates of configuration member
+	virtual void check_duplicate(const PC_tree_t& node, const std::string& name) = 0;
```
Comment (jbigot) [inline pdi/include/pdi/context.h]: I don't get what's the purpose of the function and you didn't comment the parameters (and it's not a Doxygen comment)

File: pdi/src/global_context.cxx [hunk 16]
```diff
@@ -66,15 +82,19 @@ void load_data(Context& ctx, PC_tree_t node, bool is_metadata)
 	int map_len = len(node);
 
 	for (int map_id = 0; map_id < map_len; ++map_id) {
-		Data_descriptor& dsc = ctx.desc(to_string(PC_get(node, "{%d}", map_id)).c_str());
+		std::string name = to_string(PC_get(node, "{%d}", map_id));
+
+		ctx.check_duplicate(PC_get(node, "{%d}", map_id), name);
```
Comment (jbigot) [inline pdi/src/global_context.cxx]: so you pass a node and its string representation. Isn't that kind of twice the same thing?

File: pdi/src/global_context.cxx [hunk 17]
```diff
@@ -66,15 +82,19 @@ void load_data(Context& ctx, PC_tree_t node, bool is_metadata)
...
+		Data_descriptor& dsc = ctx.desc(name.c_str());
 		dsc.metadata(is_metadata);
 		dsc.default_type(ctx.datatype(PC_get(node, "<%d>", map_id)));
 	}
-	if (is_metadata) {
+
+	if (is_metadata)
 		ctx.logger().trace("Loaded {} metadata", map_len);
-	} else {
+	else
```
Comment (jbigot) [inline pdi/src/global_context.cxx:126]: why the change? this goes against the coding guide of PDI, no?

File: pdi/docs/Specification_tree_ref.md [hunk 18]
```diff
@@ -9,28 +9,34 @@ The *specification tree root* is a **mapping** that contains the following keys:
...
+* the `logging` section specify logger properties
+* the `metadata` and `data` sections specify the type of the data in buffers
   exposed by the application; for `metadata`, %PDI keeps a copy while it only
   keeps references for `data`,
 * the `plugins` section specifies the list of plugins to load and their
   configuration,
 * the `plugin_path` section specifies the path to a directory where %PDI should
   search for plugins
-* the `logging` section specify logger properties
+* the `types` section specifies user-defined datatypes
```
Comment (jbigot) [inline pdi/docs/Specification_tree_ref.md:31]: why all the changes in order?

File: pdi/src/global_context.h [hunk 19]
```diff
@@ -73,6 +75,21 @@ class PDI_EXPORT Global_context: public Context
 
 	Global_context(Global_context&&) = delete;
 
+	// The node of each first definition, for duplicate error reporting
+	std::unordered_map<std::string, PC_tree_t> m_defined;
```
Comment (jbigot) [inline pdi/src/global_context.h]: This could be added in the descriptor, what do you think?

File: pdi/src/global_context.cxx [hunk 20]
```diff
@@ -66,15 +82,19 @@ void load_data(Context& ctx, PC_tree_t node, bool is_metadata)
 	int map_len = len(node);
 
 	for (int map_id = 0; map_id < map_len; ++map_id) {
-		Data_descriptor& dsc = ctx.desc(to_string(PC_get(node, "{%d}", map_id)).c_str());
+		std::string name = to_string(PC_get(node, "{%d}", map_id));
+
+		ctx.check_duplicate(PC_get(node, "{%d}", map_id), name);
+
+		Data_descriptor& dsc = ctx.desc(name.c_str());
```
Comment (jbigot) [inline pdi/src/global_context.cxx]: you could just check if the descriptor you get had already been filled.

File: pdi/src/global_context.cxx [hunk 21]
```diff
@@ -109,31 +129,12 @@ Global_context::Global_context(PC_tree_t conf)
...
 	// load basic datatypes
 	Datatype_template::load_basic_datatypes(*this);
-	// load user datatypes
-	Datatype_template::load_user_datatypes(*this, PC_get(conf, ".types"));
 
-	m_plugins.load_plugins();
+	load_pdi_config(conf); // single tree traversal + two flat sub-passes
 
 	// evaluate pattern after loading plugins
 	m_logger.evaluate_pattern(*this);
```
Comment (jbigot) [inline pdi/src/global_context.cxx:276]: shouldn't this be par of `load_pdi_config`?

File: pdi/src/global_context.cxx [hunk 22]
```diff
@@ -222,12 +223,186 @@ Callbacks& Global_context::callbacks()
...
+			throw Config_error(node, "Duplicate definition of '{}', originally defined in '{}', defined again in '{}'", name, redef_path, orig_path);
+		} else if (redef_has_path) {
+			throw Config_error(node, "Duplicate definition of '{}', originally defined in string config, defined again in '{}'", name, redef_path);
+		} else if (orig_has_path) {
+			throw Config_error(node, "Duplicate definition of '{}', originally defined in '{}', defined again in a string config", name, orig_path);
+		}
+#endif
+		throw Config_error(node, "Duplicate definition of '{}'", name);
+	}
+}
```
Comment (jbigot) [inline pdi/src/global_context.cxx]: Don't think you need a dedicated function

File: pdi/src/global_context.h [hunk 23]
```diff
@@ -73,6 +75,21 @@ class PDI_EXPORT Global_context: public Context
...
+	void collect_ordered_nodes_impl(
+		PC_tree_t conf,
+		std::unordered_set<std::string>& globally_loaded,
+		std::unordered_set<std::string>& include_chain,
+		std::vector<std::pair<std::string, PC_tree_t>>& ordered_nodes,
+		std::size_t& no_path_counter,
+		const std::string& known_path = "",
+		PC_tree_t include_directive = {},
+		const std::string& parent_id = ""
+	);
```
Comment (jbigot) [inline pdi/src/global_context.h]: cound't this be a static function in the cxx?

File: pdi/src/global_context.h [hunk 24]
```diff
@@ -125,8 +142,25 @@ class PDI_EXPORT Global_context: public Context
...
+	/// Diamond/circular detection is handled here.
+	void collect_ordered_nodes(
+		PC_tree_t conf,
+		std::unordered_set<std::string>& globally_loaded,
+		std::unordered_set<std::string>& include_chain,
+		std::vector<std::pair<std::string, PC_tree_t>>& ordered_nodes,
+		const std::string& known_path = "",
+		PC_tree_t include_directive = {},
+		const std::string& parent_id = ""
+	);
```
Comment (jbigot) [inline pdi/src/global_context.h]: cound't this be a static function in the cxx?

**Review by jbigot — CHANGES_REQUESTED** [review]

File: pdi/docs/Specification_tree_ref.md (hunk 18 above)
Comment (JAuriac) [reply pdi/docs/Specification_tree_ref.md:31]: The previous documentation order did not match its summary

File: pdi/src/global_context.cxx (hunk 22 above)
Comment (JAuriac) [reply pdi/src/global_context.cxx]: Can we get the 'Config_error' with the lines numbers without a new function ?

File: pdi/src/global_context.cxx (hunk 21 above)
Comment (JAuriac) [reply pdi/src/global_context.cxx:276]: I kept this separated as it previously was, to not make a unique monolithic function inside of the Global_context definition

File: pdi/src/global_context.h (hunk 23 above)
Comment (JAuriac) [reply pdi/src/global_context.h]: 'collect_ordered_nodes_impl', yes, but not 'collect_ordered_nodes', as we need a public member, used by 'load_pdi_config' ?

File: pdi/src/global_context.h (hunk 24 above)
Comment (JAuriac) [reply pdi/src/global_context.h]: 'collect_ordered_nodes_impl', yes, but not 'collect_ordered_nodes', as we need a public member, used by 'load_pdi_config' ?

File: pdi/src/global_context.cxx (hunk 13 above)
Comment (JAuriac) [reply pdi/src/global_context.cxx:80]: Changed again, to use make_and_check_descriptor() which is accessible only from Global_context (and not Context)

File: pdi/src/global_context.h (hunk 24 above)
Comment (JAuriac) [reply pdi/src/global_context.h]: Moved both collect_ordered_nodes() and collect_ordered_nodes() to the local namespace instead

File: pdi/src/global_context.h (hunk 23 above)
Comment (JAuriac) [reply pdi/src/global_context.h]: Moved both collect_ordered_nodes() and collect_ordered_nodes() to the local namespace instead

**Comment by jbigot** [comment]
Replaced by #704
